pygfx/objects/_base.py

```diff
@@ -306,63 +294,92 @@ def receive_shadow(self, value):
-        """
-        return tuple(self._children)
-
-    def add(self, *objects, before=None):
-        """Adds object as child of this object. Any number of
-        objects may be added. Any current parent on an object passed
-        in here will be removed, since an object can have at most one
-        parent.
-        If ``before`` argument is given (and present in children), then
-        the items are inserted before the given element.
         """
-        idx = len(self._children)
-        if before:
-            try:
-                idx = self._children.index(before)
-            except ValueError:
-                pass
+
+        obj: WorldObject
         for obj in objects:
-            assert isinstance(obj, WorldObject)
-            # orphan if needed
-            if obj._parent_ref is not None:
-                obj._parent_ref().remove(obj)
-            # attach to scene graph
-            obj._parent_ref = weakref.ref(self)
-            self._children.insert(idx, obj)
-            idx += 1
-            # flag world matrix as dirty
-            obj._matrix_world_dirty = True
+            if obj.parent is not None:
+                obj.parent.remove(obj, keep_world_matrix=keep_world_matrix)
+
+            if before is not None:
+                idx = self.children.index(before)
+            else:
+                idx = len(self.children)
+
+            if keep_world_matrix:
+                transform_matrix = obj.world.matrix
+
+            obj._parent = weakref.ref(self)
+            obj.world.parent = self.world
+            self.children.insert(idx, obj)
+
+            if keep_world_matrix:
+                obj.world.matrix = transform_matrix
+
         return self
 
-    def remove(self, *objects):
-        """Removes object as child of this object. Any number of objects may be removed.
-        If a given object is not a child, it is ignored.
-        """
+    def remove(self, *objects, keep_world_matrix=False):
+        """Removes object as child of this object. Any number of objects may be removed."""
+
+        obj: WorldObject
         for obj in objects:
             try:
-                self._children.remove(obj)
-                obj._parent_ref = None
+                self.children.remove(obj)
             except ValueError:
-                pass
-        return self
+                warnings.warn(
+                    "Attempting to remove object that was not a child.", UserWarning
+                )
+                continue
+            else:
+                obj._reset_parent(keep_world_matrix=keep_world_matrix)
 
-    def clear(self):
+    def clear(self, *, keep_world_matrix=False):
         """Removes all children."""
-        for child in self._children:
-            child._parent_ref = None
-        self._children.clear()
+
+        for child in self.children:
+            child._reset_parent(keep_world_matrix=keep_world_matrix)
+
+        self.children = []
```

Comment:
`.clear()` puts less pressure on the GC than allocating a new object :)

Reply:
Sure.